lib/safetensors.ex

```diff
@@ -85,6 +85,43 @@ defmodule Safetensors do
     ]
   end
 
+  @doc """
+  Reads a safe tensor from file.
+
+  Tensors are loaded into Nx one by one,
+  without loading the whole file into disk.
+  """
+  @spec read!(path :: iodata()) :: %{String.t() => Nx.Tensor.t()}
+  def read!(path) do
+    File.open!(path, [:read, :raw], fn file ->
+      {:ok, <<header_size::unsigned-64-integer-little>>} = :file.read(file, 8)
+      {:ok, header_json} = :file.read(file, header_size)
+
+      header = decode_header!(header_json)
+
+      for {tensor_name, tensor_info} <- header, into: %{} do
+        %{
+          "data_offsets" => [offset_start, offset_end],
+          "dtype" => dtype,
+          "shape" => shape
+        } = tensor_info
+
+        {_, elem_size} = type = dtype_to_type(dtype)
+
+        {:ok, binary} =
+          :file.pread(file, header_size + 8 + offset_start, offset_end - offset_start)
+
+        tensor =
+          binary
+          |> new_byte_order(elem_size, :little)
+          |> Nx.from_binary(type)
+          |> Nx.reshape(List.to_tuple(shape))
```

Comment:
We can share more with the other function, wdyt?

```elixir
defp build_tensor(binary, tensor_info) do
  %{"dtype" => dtype, "shape" => shape} = tensor_info
  {_, elem_size} = type = dtype_to_type(dtype)

  binary
  |> new_byte_order(elem_size, :little)
  |> Nx.from_binary(type)
  |> Nx.reshape(List.to_tuple(shape))
end
```

Reply:
Good call, done!